kernel/device.c

```diff
@@ -79,10 +80,7 @@ const struct device *z_impl_device_get_binding(const char *name)
 {
 	const struct device *dev;
 
-	/* A null string identifies no device.  So does an empty
-	 * string.
-	 */
-	if ((name == NULL) || (*name == 0)) {
+	CHECKIF((name == NULL) || (*name == 0)) {
```

Comment:
No, this is wrong.  It is expected now to get strings that are empty; that piece cannot be conditional on build options.

Reply:
Why it is expected *now* ? From my understanding you are passing a wrong parameter to this api and is expecting it to handle it properly and not crash, that  behavior is completely different from other APIs isn't ? The question is why only this API ? You know that if we start checking parameters in implementation functions we will end up with a lot of redundant checks between verification and implementation functions when user mode is enabled. We will need to properly review how this is done if that is the path.

Reply:
You're lacking context.  Please, look at #32266.  

To recap: We are starting to remove `label` properties from devicetree, but devices still have a `name` field.  Rather than make that NULL, which saves the most space but would cause faults in a lot of log messages that pass `dev->name` to `%s`, we've decided to set the `name` to an empty string when no name is available.

Elimination of required `label` means that `device_get_binding()` needs to behave reasonably when a string that ultimately derives from `dev->name` is passed in.  If `dev->name` is an empty string it can't be used to identify the device, so it can't be guaranteed to return `dev`.  Instead the function should return immediately indicating that no device could be found, rather than picking the first device that didn't have a name (as would happen if the check wasn't performed).

It also made sense to have it behave reasonably when a null pointer was passed in, but I understand you don't like that.

Reply:
ok, device name can be null now, will close this pr, the behavior has changed.